modules/dists/SparseBlockDist.chpl

```diff
@@ -117,6 +119,102 @@ class SparseBlockDom: BaseSparseDom {
+      preserveInds=true){
+  
+    if !isSorted && preserveInds {
+      var _inds = inds;
+      bulkAdd_help(_inds, isSorted, isUnique); 
+    }
+    else {
+      bulkAdd_help(inds, isSorted, isUnique);
+    }
+  }
+
+  // Tried to put this record in the function and the if statement, but got a
+  // segfault from the compiler.
+  record TargetLocaleComparator {
+
+    // I wish I could return tuple from key function. There are checks to see
+    // if the type of the function is numeric/string. Maybe there should be
+    // isComparable that can do recursive checks for at least tuples of
+    // numericals/strings/comparables.
+    // In general being able to return tuples from key() can enable sorting with
+    // multiple keys.
+
+    //proc key(a: index(rank, idxType)) { 
+    //  return dist.targetLocales[dist.targetLocsIdx(a)].id; 
+    //}
+
+    proc compare(a: index(rank, idxType), b: index(rank, idxType)) {
+      var locDif = dist.targetLocales[dist.targetLocsIdx(a)].id -
+        dist .targetLocales[dist.targetLocsIdx(b)].id;
+
+      if locDif != 0 then return locDif;
+
+      if a < b then return -1;
+      if a > b then return 1;
+      return 0;
+    }
+  }
+
+  proc bulkAdd_help(inds: [] index(rank,idxType), isSorted=false, isUnique=false) {
+
+    if !isSorted {
+
+      // without _new_ krecord functions throw null deref. It doesn't seem to be 
+      // able to deref _outer_ ? I think it has something to do with memory
+      // allocation for classes vs records, but I cannot explain
+      var comp = new TargetLocaleComparator;
+      sort(inds, comparator=comp);
+    }
+
+    var localIndexInitSize = max(1, inds.size/numLocales);
+    var localIndexDom = {0..#localIndexInitSize};
+
+    var localIndexes: [dist.targetLocDom] [localIndexDom] index(rank,idxType);
+    var localIndexCnts: [dist.targetLocDom] int =
+      inds.domain.low;
+
+    if bulkAddMemoize {
+      var lastLocaleIndexCache = dist.targetLocDom.low;
+      for i in inds {
+        if locDoms[lastLocaleIndexCache].parentDom.member(i) {
+          addToLocalIndexes(i,lastLocaleIndexCache);
+        }
+        else {
+          const localeIndex = dist.targetLocsIdx(i);
+          addToLocalIndexes(i, localeIndex);
+        }
+      }
+    }
+    else {
+      for i in inds {
+        const localeIndex = dist.targetLocsIdx(i);
+        addToLocalIndexes(i, localeIndex);
+      }
+    }
+    coforall localeIdx in dist.targetLocDom do {
+      on dist.targetLocales(localeIdx) do {
+        const locCnt = localIndexCnts[localeIdx];
+        locDoms[localeIdx].mySparseBlock.bulkAdd(localIndexes[localeIdx][0..#locCnt], 
+            true, false, false); 
+      }
+    }
+
+    // helper to grow localIndexes array if necessary
+    proc addToLocalIndexes(idx, locId) {
+
+      if localIndexes[locId].size <= localIndexCnts[locId] {
+        localIndexDom = {0..#localIndexes[locId].size*2};
+      }
```

Comment:
Did you consider using push_back and not do it for some reason? Also, it seems to me if you keep this code here, it would be useful to bound localIndexDom.hi by inds.size.